support-lib/cpp/Future.hpp

```diff
@@ -0,0 +1,193 @@
+        setException(std::make_exception_ptr(ex));
+    }
+
+    void setException(std::exception_ptr ex) {
+        SharedStatePtr<T> sharedState;
+        {
+            std::unique_lock lk(_sharedState->mutex);
+            _sharedState->exception = ex;
+            sharedState = std::move(_sharedState);
+        }
+        callResultHandler(sharedState);
+    }
+private:
+    SharedStatePtr<T> _sharedState = std::make_shared<SharedState<T>>();
+
+    void callResultHandler(SharedStatePtr<T> sharedState) {
+        if (sharedState->handler) {
+            // handler already assigned
+            sharedState->handler(sharedState);
+        } else {
+            sharedState->cv.notify_all();
+        }
+    }
+};
+
+template <typename T>
+class Promise: public PromiseBase<T> {
+public:
+    using PromiseBase<T>::setValue;
+    using PromiseBase<T>::setException;
+};
+
+template <>
+class Promise<void>: public PromiseBase<void> {
+public:
+    void setValue() {
+        PromiseBase<void>::setValue(true);
+    }
+    using PromiseBase<void>::setException;
+};
+
+template<typename T>
+class Future {
+    template<typename U>
+    friend class PromiseBase;
+    
+    Future(SharedStatePtr<T> sharedState) : _sharedState(sharedState) {}
+public:
+    Future(const Future& other) = default;
+    Future(Future&& other) noexcept = default;
+    Future& operator= (Future&& other) noexcept = default;
+
+    bool isValid() const {
+        return _sharedState != nullptr;
+    }
+    
+    bool isReady() const {
+        std::unique_lock lk(_sharedState->mutex);
+        return _sharedState->isReady();
+    }
+    
+    void wait() const {
+        std::unique_lock lk(_sharedState->mutex);
+        _sharedState->cv.wait(lk, [state = _sharedState] {return state->isReady();});
+    }
+
+    auto get() const {
+        std::unique_lock lk(_sharedState->mutex);
+        _sharedState->cv.wait(lk, [state = _sharedState] {return state->isReady();});
+        if (!_sharedState->exception) {
+            return _sharedState->getValueUnsafe();
+        } else {
+            std::rethrow_exception(_sharedState->exception);
+        }
+    }
+
+    template<typename FUNC>
+    auto then(FUNC handler) {
+        using HandlerReturnType = std::invoke_result_t<FUNC, Future<T>>;
+
+        SharedStatePtr<T> sharedState;
+        {
+            std::unique_lock lk(_sharedState->mutex);
+            sharedState = std::move(_sharedState);
+        }
+        Promise<HandlerReturnType> nextPromise;
+        auto nextFuture = nextPromise.getFuture();
+        auto continuation = [this, handler, nextPromise] (SharedStatePtr<T> x) mutable {
+            try {
+                if constexpr(std::is_void_v<HandlerReturnType>) {
+                    handler(Future<T>(x));
+                    nextPromise.setValue();
+                } else {
+                    nextPromise.setValue(handler(Future<T>(x)));
+                }
+            } catch (std::exception& e) {
```

Comment:
can we disable the try/catch in the event -fno-exceptions is set?

Reply:
Let's do this later.  Right now there are so many try/catches in djinni and its generated code there's no point to attempt no-exceptions here.